test/e2e-nokey/upload.e2e.test.js

```diff
@@ -0,0 +1,379 @@
+
+      // Submit form and wait for response
+      const [response] = await Promise.all([page.waitForResponse((response) => response.url().includes('/api/upload') && response.request().method() === 'POST'), page.click('button[type="submit"]')]);
+
+      // Verify redirect response
+      expect(response.status()).toBe(302);
+
+      // Wait for redirect to complete
+      await page.waitForURL('/api/upload');
+      await page.waitForLoadState('networkidle');
+
+      // Check for error message about file size
+      const errorAlert = page.locator('.alert.alert-danger');
+      await expect(errorAlert).toBeVisible({ timeout: 10000 });
+      await expect(errorAlert).toContainText(/file size.*too large.*1MB/i);
+    } finally {
+      // Clean up test file
+      if (fs.existsSync(largeFilePath)) {
+        fs.unlinkSync(largeFilePath);
+      }
+    }
+  });
+
+  test('should handle form submission without file', async ({ page }) => {
+    await page.goto('/api/upload');
+    await page.waitForLoadState('networkidle');
+
+    // Verify CSRF token is present
+    const csrfInput = page.locator('input[name="_csrf"]');
+    await expect(csrfInput).toBeAttached();
+
+    // Submit form and wait for response
+    const [response] = await Promise.all([page.waitForResponse((response) => response.url().includes('/api/upload') && response.request().method() === 'POST'), page.click('button[type="submit"]')]);
+
+    // Verify redirect response
+    expect(response.status()).toBe(302);
+
+    // Wait for redirect to complete
+    await page.waitForURL('/api/upload');
+    await page.waitForLoadState('networkidle');
+
+    // Should redirect back to upload page (no file selected is handled gracefully)
+    await expect(page.locator('h2')).toContainText('File Upload');
+  });
+
+  test('should maintain CSRF protection', async ({ page }) => {
+    await page.goto('/api/upload');
+    await page.waitForLoadState('networkidle');
+
+    // Verify CSRF token is present
+    const csrfInput = page.locator('input[name="_csrf"]');
+    await expect(csrfInput).toBeAttached(); // CSRF is hidden, so check for presence
+
+    const csrfValue = await csrfInput.getAttribute('value');
+    expect(csrfValue).toBeTruthy();
+    expect(csrfValue.length).toBeGreaterThan(0);
+
+    // Verify form has correct enctype for file uploads
+    const form = page.locator('form[enctype="multipart/form-data"]');
+    await expect(form).toBeVisible();
+    await expect(form).toHaveAttribute('method', 'POST');
+  });
+
+  test('should handle upload with maximum allowed file size', async ({ page }) => {
+    await page.goto('/api/upload');
+    await page.waitForLoadState('networkidle');
+
+    // Create tmp directory if it doesn't exist
+    const tmpDir = path.join(__dirname, '../../tmp');
+    if (!fs.existsSync(tmpDir)) {
+      fs.mkdirSync(tmpDir, { recursive: true });
+    }
+
+    // Create a file exactly at the 1MB limit (1024 * 1024 bytes)
```

Comment:
Probably reword `// Create a file exactly at the 1MB limit (1024 * 1024 bytes)` so it matches the comment on the line right after it`